### PR TITLE
Feature/binned multiproc duration extraction

### PR DESCRIPTION
**Pull request by cschaefer26**
- Much quicker duration extraction with equal precision with flexible binning and multiprocessing

### CODE REVIEW

File: duration_extraction/duration_extraction_pipe.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        """
+
+        device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
+        model.to(device)
+
+        dataloader = get_binned_taco_dataloader(data_path=self.paths.data, max_batch_size=max_batch_size)
+
+        sum_items = 0
+        sum_att_score = 0
+        pbar = tqdm(dataloader, total=len(dataloader), smoothing=0.01)
```
Comment (alexteua) [inline duration_extraction/duration_extraction_pipe.py:111]: nice using smoothing parameter, I somehow also missed till now that it exists :)

File: utils/dataset.py [hunk 2]
```diff
@@ -131,6 +132,146 @@ def __call__(self, batch: List[Dict[str, torch.tensor]]) -> Dict[str, torch.tens
...
+        consecutive_split_points = np.where(np.diff(dataset_lens, append=0, prepend=0) != 0)[0]
+        dataset_indices = list(range(len(dataset)))
+        all_batches = []
+
+        for a, b in zip(consecutive_split_points[:-1], consecutive_split_points[1:]):
+            big_batch = dataset_indices[a:b]
+            batches = list(batchify(big_batch, batch_size=max_batch_size))
+            all_batches.extend(batches)
+
+        Random(42).shuffle(all_batches)
```
Comment (alexteua) [inline utils/dataset.py:260]: would be nice to move seed value to some constant or config entry

File: train_tacotron.py [hunk 3]
```diff
@@ -101,57 +101,35 @@ def create_gta_features(model: Tacotron,
...
+    duration_extractor = DurationExtractor(silence_threshold=durex_conf['silence_threshold'],
+                                           silence_prob_shift=durex_conf['silence_prob_shift'])
+
+    duration_extraction_pipe = DurationExtractionPipeline(paths=paths, config=config,
+                                                          duration_extractor=duration_extractor)
+
+    print('Extracting attention matrices from tacotron...')
+    #duration_extraction_pipe.extract_attentions(model, max_batch_size=durex_conf['max_batch_size'])
 
+    num_workers = durex_conf['num_workers']
```
Comment (alexteua) [inline train_tacotron.py:123]: I would rename "durex_conf" as it reminds of a certain brand :)

**Review by alexteua — CHANGES_REQUESTED** [review]
Really nice implementation that should speed things up!
Just a minor comments, plus I would suggest to give more explicit names to variables, for example att -> attention, dur -> duration, sig -> ? etc

File: utils/dataset.py (hunk 2 above)
Comment (cschaefer26) [reply utils/dataset.py:260]: Moved to a constant. IMO not necessary to allow the setting in a config as the shuffling is merely used for a better time estimation of the inference.

File: train_tacotron.py (hunk 3 above)
Comment (cschaefer26) [reply train_tacotron.py:123]: Lol, renamed.

**Comment by alexteua** [comment]
@cschaefer26 I would still suggest to improve variables names by making them more explicit for better code readability (for example att -> attention, dur -> duration, sig -> ? etc)

**Comment by cschaefer26** [comment]
> @cschaefer26 I would still suggest to improve variables names by making them more explicit for better code readability (for example att -> attention, dur -> duration, sig -> ? etc)

Hi, could you be a bit more specific? I would not like to expand all abbreviated names as it would make the code less readable due to variable lenght (e.g. att_score_dict -> attention_score_dictionary is too long in my opinion)

**Comment by alexteua** [comment]
> > @cschaefer26 I would still suggest to improve variables names by making them more explicit for better code readability (for example att -> attention, dur -> duration, sig -> ? etc)
> 
> Hi, could you be a bit more specific? I would not like to expand all abbreviated names as it would make the code less readable due to variable lenght (e.g. att_score_dict -> attention_score_dictionary is too long in my opinion)

@cschaefer26  sure, I just mentioned above the ones that I noticed in multiple variable names. For example:
- att which can be expanded to attention
- dur expanded to duration
- sig -> I don't know what that abbreviation means

with the example of att_score_dict I think attention_score_dict would be not too long but also understandable. 

I understand the concern with readability of too long variables, but using too shorted and abbreviated ones reduces readability too

**Review by alexteua — APPROVED** [review]
